Feat/Add 3scale high availability documentation (#15)
3scale/3scale-Operations: docs/ha_3scale.adoc

:toc:
:toc-placement!:

= 3scale HA different scenarios

toc::[]

== Common configuration

* This section explains how to install 3scale using the link:https://github.com/3scale/3scale-operator[3scale Operator] and how to configure it to have a highly-available deployment
* The first thing that you need to do is install the operator following  link:https://github.com/3scale/3scale-operator/blob/3scale-2.10-stable/doc/operator-user-guide.md#installing-3scale[this guide]

> * After installing the operator, you need to deploy an `APIManager` custom resource, that will be read by the operator and then deploy 3scale configured according to its spec

Reviewer: ```suggestion
* With the 3scale-operator installed and ready in the cluster, deploy an `APIManager` custom resource. The operator will install 3scale in the cluster as instructed by the `APIManager` spec.
```

> * There is a basic example of an `APIManager` link:https://github.com/3scale/3scale-operator/blob/3scale-2.10-stable/doc/operator-user-guide.md#basic-installation[here]

Reviewer: ```suggestion
* There is a basic example of an `APIManager` custom resource link:https://github.com/3scale/3scale-operator/blob/3scale-2.10-stable/doc/operator-user-guide.md#basic-installation[here]
```

> * In order to set up an HA deployment you need to tune a few things in the `APIManager` spec for every component

Reviewer: ```suggestion
* In order to set up an HA deployment you need to tune a few things in the `APIManager` custom resource spec for every component
```

=== External databases

> * You need to tell the Operator not to deploy the critical DBs following link:ha_dbs.adoc[this 3scale HA databases section] of this repo

Reviewer: ```suggestion
* You need to instruct the operator not to deploy the critical databases of the 3scale installation. You can check the link:ha_dbs.adoc[3scale HA databases documentation section] for this.
```

> * Notice that the operator does not offer an option to deploy `system-memcached` and `system-sphinx` externally *because those two are not considered critical*, so you do not need an HA deployment of them

Reviewer: ```suggestion
* Notice that the operator does not offer an option to deploy `system-memcached` or `system-sphinx` externally. Neither of these components have a critical impact on 3scale in the event they are down, so the operator does not currently support a high-available deployment option for them.
```

Saying *you don't need* and HA deployment for them is a bit an extreme/strong statement.

=== Scaling number of replicas

> * You need to deploy at least 2 replicas of each component that can scale. That includes all the pods except the DBs

Reviewer: This needs rewording, it's a very confusing statement. If I were a reader with no beforehand knowledge of 3scale would read this and think:

* "2 replicas of each component that can scale": well of course, that's why I'm reading this, but I don't know which ones can scale
* "That includes all the pods except the DBs": wait what? which databases? I thought databases were external ... (puzzled face)

Author: Fixed https://github.com/3scale/3scale-Operations/commit/080369724359e1f2dd29c0c4665236d0e1100601

=== Scaling number of replicas

The number of replicas of each component needs to be at least 2 to support pod failures. For installations with nodes running in different availability zones, the minimum replica count should match the number of availability zones available. This configuration used in combination with node-based pod anti-affinity rules, described in the next section, will ensure that at least one pod pod will run in each availability zone. 3scale components that support more than one replicas are:

- apicast-production
- apicast-staging
- backend-listener
- backend-worker
- system-app
- system-sidekiq
- zync
- zync-que

Depending on your cluster size, throughput and latency requirements you might want to deploy more replicas, at least for the components that are called on each request to the APIs configured in 3scale, that is, `apicast-production`, `backend-listener`, and `backend-worker`.

Replicas can be configured using the `replicas` attribute in the `APIManager` spec.

=== Pod Affinity

> Enabling pod affinities in the operator for every component, ensures that pod replicas from each deployment are distributed across different nodes of the cluster, and even balanced across different availability zones. You can read more about affinities in the link:https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity[Kubernetes docs].

Reviewer: ```suggestion
Enabling pod affinities in the operator for every component, ensures that pod replicas from each deployment are distributed across different nodes of the cluster, and evenly balanced across different availability zones. You can read more about affinities in the link:https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity[Kubernetes docs].
```

To enable affinities you need to follow link:https://github.com/3scale/3scale-operator/blob/master/doc/operator-user-guide.md#setting-custom-affinity-and-tolerations[these examples].

> The following example to be added for example to `apicastProductionSpec` (but can be added to any non-database deployment), adds a *soft podAntiAffinity* configuration using `preferredDuringSchedulingIgnoredDuringExecution` (so scheduler will try to run this set of `apicast-production` pods in different hosts from different availability zones, but if it is not possible, then allow to run elsewhere):

Reviewer: ```suggestion
The following affinity block can be added for example to `apicastProductionSpec` (but can be added to any non-database DeploymentConfig), adds a *soft podAntiAffinity* configuration using `preferredDuringSchedulingIgnoredDuringExecution` (so the scheduler will try to run this set of `apicast-production` pods in different hosts from different availability zones, but if it is not possible, then allow to run elsewhere):
```

Author: fixed

```yaml
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    deploymentConfig: apicast-production
                topologyKey: kubernetes.io/hostname
            - weight: 99
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    deploymentConfig: apicast-production
                topologyKey: topology.kubernetes.io/zone
```

> In the following example, unlike previous one, it sets a *hard podAntiAffinity* configuration using `requiredDuringSchedulingIgnoredDuringExecution` (so conditions must be met for a pod to be scheduled onto a node, otherwise pod won’t be scheduled, which can be risky on certain situations like being on a cluster with low free resources, and new pods cannot be scheduled):

Reviewer: ```suggestion
In the following example, unlike in the previous one, a *hard podAntiAffinity* configuration is set using `requiredDuringSchedulingIgnoredDuringExecution` (so conditions must be met for a pod to be scheduled onto a node, otherwise pod won’t be scheduled, which can be risky on certain situations like being on a cluster with low free resources, and new pods cannot be scheduled):
```

Author: fixed

```yaml
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    deploymentConfig: apicast-production
                topologyKey: kubernetes.io/hostname
            - weight: 99
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    deploymentConfig: apicast-production
                topologyKey: topology.kubernetes.io/zone
```

=== Pod Disruption Budgets

Enabling pod disruption budgets (PDBs) ensures that a minimum number of pods will be available for each component. You can read more about PDBs in the link:https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets[Kubernetes docs].

Pod disruption budgets configuration is documented in link:https://github.com/3scale/3scale-operator/blob/3scale-2.10-stable/doc/operator-user-guide.md#enabling-pod-disruption-budgets[the user guide].

'''

> The information provided above enables a basic highly-available installation, , but you might need to take some extra steps depending on the environment where 3scale is running. Next, there are some of the most common scenarios described.

Reviewer: ```suggestion
The information provided above enables a basic highly-available installation, but you might need to take some extra steps depending on the environment where 3scale is running. Next, there are some of the most common scenarios described.
```

Author: fixed

== Single cluster in single availability zone

With this setup, 3scale will continue working if a node fails, but stop working in case the availability zone fails.

Pod affinities do apply (but only using one rule with `kubernetes.io/hostname`), because there is a single availability zone.

Pod disruption budgets do apply.

== Single cluster in multiple availability zones

The same setup as before, but now nodes are distributed over hosts from different availability zones (physically separated locations or data centers within a cloud region that are tolerant to local failures).

A minimum of 3 availability zones is recommended to have high availability.

With this setup, 3scale will continue working if a node or even an availability zone fails at the same time.

Pod affinities do apply (with both rules `kubernetes.io/hostname` and `topology.kubernetes.io/zone`).

Pod disruption budgets do apply.

== Multiple clusters in multiple availability zones

> There are several options to install 3scaleacross several OpenShift clusters and availability zones.

Reviewer: ```suggestion
There are several options to install 3scale across several OpenShift clusters and availability zones.
```

In multiple cluster installation options, clusters work in an active/passive configuration, with the failover procedure involving a few manual steps. Note that there will be service disruption while a human operator performs the required steps to bring the passive cluster into active mode in case of failure.

> This documentation focuses on deployment using Amazon Web Services (AWS), but the same patterns and configuration options should apply to other public cloud vendors as long as the provider’s managed DB services offer the required features (multi-AZ, multi-region…​).

Reviewer: ```suggestion
This documentation focuses on deployment using Amazon Web Services (AWS), but the same patterns and configuration options should apply to other public cloud vendors as long as the provider’s managed DB services offer the required features (support for multiple availability zones, multiple regions …​).
```

Author: fixed

* The secrets containing credentials like tokens, passwords, etc, have to be manually deployed in all clusters with the same values. By default, the 3scale operator creates them with some secure random values on every cluster. However, in this case, you need to have the same credentials in both clusters. The list of secrets and how to configure them can be found in the link:https://github.com/3scale/3scale-operator/blob/master/doc/apimanager-reference.md#apimanager-secrets[3scale Operator docs]. This is the list of secrets that should be mirrored in both clusters:
- backend-internal-api
- system-app
- system-events-hook
- system-master-apicast
- system-seed
* The secrets containing DB connections strings (`backend-redis`, `system-database`, `system-redis`, `zync`) have to be manually deployed like explained on link:ha_dbs.adoc[external databases]:
- If databases are shared among clusters they must use the same values on all clusters

> - While if each cluster have their own databases, they must use different values on each cluster

Reviewer: ```suggestion
- On the other hand, if each cluster have their own databases, they must use different values on each cluster
```

Author: fixed

=== Active-Passive clusters on the same region with shared databases

This setup consists of having 2 clusters (or more) in the *same region* and deploying 3scale in *active-passive* mode. One of the clusters will be the *active* (receiving traffic), whereas the other/s will be in standby mode without receiving traffic (*passive*), but prepared to assume the active role in case there is a failure in the active cluster.

> In this installation option, being a single region involved, databases will be shared among all clusters.

Reviewer: ```suggestion
In this installation option, given that only a single region is used, databases will be shared among all clusters.
```

Author: fixed
